### PR TITLE
Output

### PR DESCRIPTION
**Pull request by TApplencourt**
Added `--trace-output` and `--analysis-output`.  (to save the trace, and the output of the analysis into specific file)
Added also a `--timeline [PATH]` so one can change where the timeline will be stored

```
applenco@x1921c0s6b0n0:~/THAPI/build> ./ici/bin/iprof  -l roger.pftrace  sycl-ls
Warning: ONEAPI_DEVICE_SELECTOR environment variable is set to level_zero:gpu.
To see the correct device id, please unset ONEAPI_DEVICE_SELECTOR.

[ext_oneapi_level_zero:gpu:0] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:1] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:2] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:3] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:4] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:5] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
THAPI: Trace location: /home/applenco/thapi-traces/thapi_interval--2024-02-19--21h52m08s
THAPI: Perfetto trace location: roger.pftrace
applenco@x1921c0s6b0n0:~/THAPI/build> ./ici/bin/iprof  -l -- sycl-ls
Warning: ONEAPI_DEVICE_SELECTOR environment variable is set to level_zero:gpu.
To see the correct device id, please unset ONEAPI_DEVICE_SELECTOR.

[ext_oneapi_level_zero:gpu:0] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:1] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:2] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:3] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:4] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
[ext_oneapi_level_zero:gpu:5] Intel(R) Level-Zero, Intel(R) Data Center GPU Max 1550 1.3 [1.3.26241]
THAPI: Trace location: /home/applenco/thapi-traces/thapi_interval--2024-02-19--21h52m23s
THAPI: Perfetto trace location: out.pftrace
```

Note the `--` so `--timline` don't take `sycl-ls` as output.

### CODE REVIEW

**Comment by TApplencourt** [comment]
Added `thapi_metadata.yaml`. A little more viscous than expect because of the on node processing (`THAPI_DIR/$HOSTNAME`) 
First pass, need to make it cleaner now.

File: utils/babeltrace_thapi.in [hunk 1]
```diff
@@ -250,13 +254,47 @@ end
...
+  end
+  File.write(File.join($options[:output], THAPI_METADATA_FILE), y.to_yaml)
+end
+
+def raise_if_command_unvalid(command, trace)
+  case command
+  when 'trace'
+    raise('Cannot show the trace of an interval or aggregated trace') if %w[interval aggreg].include?(thapi_metadata(trace)[:type])
+  when 'timeline'
+    raise('Cannot show the trace of an aggreg trace') if %w[aggreg].include?(thapi_metadata(trace)[:type])
```
Comment (Kerilk) [inline utils/babeltrace_thapi.in:283]: Error string seems strange shouldn't it say something about timeline?

File: utils/babeltrace_thapi.in [hunk 2]
```diff
@@ -340,8 +379,10 @@ end
 # Fix segfault
 ARGV.uniq!
 
-unless $options[:muxer]
-  # The first filter of the grap to read multiple port (aka metababael plugin)
+raise_if_command_unvalid(command, ARGV.first)
```
Comment (Kerilk) [inline utils/babeltrace_thapi.in:382]: unvalid => invalid ?

File: xprof/xprof.rb.in [hunk 3]
```diff
@@ -146,7 +147,7 @@ end
 #     Each nodes need to aggree on a root folder to write into.
 #  - thapi_trace_dir_root
 #     Final directory exposed to the user. The `master rank`  rename the `thapi_trace_dir_tmp`
-#     with a more "friendly" name.
+#     with a more "friendly" name. No required if `analsys-output` is set
```
Comment (Kerilk) [inline xprof/xprof.rb.in:150]: No => Not?

File: xprof/xprof.rb.in [hunk 4]
```diff
@@ -462,28 +469,33 @@ end
 #
 
 # TODO: Use babeltrace_thapi as a LIB not a binary
-
+# Note: I don't understand `mv`
+#       File.mv(a, b) will put a into b
+#       FileUtils.rename(a,b) will move a as b, but will may
+#         Invalid cross-device
```
Comment (Kerilk) [inline xprof/xprof.rb.in:475]: I don't understand the comment

File: xprof/xprof.rb.in [hunk 5]
```diff
@@ -574,27 +591,37 @@ end
...
 end
 
 # Avoid load problem
-if __FILE__ == $0
+if __FILE__ == $PROGRAM_NAME
   parser = OptionParserWithDefaultAndValidation.new
 
+  parser.on('--trace-output PATH', 'Define where the ctf trace will be saved',
+            'Default path is something like: $THAPI_HOME/thapi-traces/thapi--%Y-%m-%d--%Hh%Mm%Ss',
+            '$THAPI_HOME is an env variable, is not set will default to $HOME') do |p|
```
Comment (Kerilk) [inline xprof/xprof.rb.in:603]: is not set => if not set?

File: xprof/xprof.rb.in (hunk 3 above)
Comment (TApplencourt) [reply xprof/xprof.rb.in:150]: Fixed with `b61cd06`

File: xprof/xprof.rb.in (hunk 4 above)
Comment (TApplencourt) [reply xprof/xprof.rb.in]: Fixed with `927531f`

**Review by Kerilk — APPROVED** [review]

File: xprof/xprof.rb.in [hunk 6]
```diff
@@ -574,27 +591,37 @@ end
...
 end
 
 # Avoid load problem
-if __FILE__ == $0
+if __FILE__ == $PROGRAM_NAME
   parser = OptionParserWithDefaultAndValidation.new
 
+  parser.on('--trace-output PATH', 'Define where the ctf trace will be saved',
+            'Default path is something like: $THAPI_HOME/thapi-traces/thapi--%Y-%m-%d--%Hh%Mm%Ss',
+            '$THAPI_HOME default to $HOME') do |p|
```
[the remark this answers is not in the document]
Comment (TApplencourt) [reply xprof/xprof.rb.in]: Why outdated ><

[the remark this answers is not in the document]
Comment (TApplencourt) [reply xprof/xprof.rb.in]: because merge by you \o/

[the remark this answers is not in the document]
Comment (Kerilk) [reply xprof/xprof.rb.in]: yes

**Pull request merged**